### ISSUE TITLE
Segmentation Fault after google-api-core upgrade

### ISSUE THREAD

**Issue by sstoops** [DataDog/dd-trace-py#8648]
### Summary of problem

This issue is a continuation of the [comments](https://github.com/DataDog/dd-trace-py/issues/8504#issuecomment-1965148147) I'd left on a similar issue.

We recently started getting segmentation faults across our environments after upgrading to version `2.17.0`. Installing every commit of `google-api-core` between `2.16.2` and `2.17.0` narrowed the issue down to [this specific commit](https://github.com/googleapis/python-api-core/commit/94726e739698035b00667983f854c600252abd28).

After much testing, I discovered that the issue would go away if I disabled ddtrace's `grpc` module (`DD_PATCH_MODULES=grpc:false`).

### Which version of dd-trace-py are you using?

`2.7.1`

### Which version of pip are you using?
```
❯ pip --version
pip 24.0 from /usr/local/lib/python3.12/site-packages/pip (python 3.12)
```

### Which libraries and their versions are you using?

<details>
  <summary>`pip freeze`</summary>

```
attrs==23.2.0
bytecode==0.15.1
cachetools==5.3.3
cattrs==23.2.3
certifi==2024.2.2
charset-normalizer==3.3.2
ddsketch==2.0.4
ddtrace==2.7.1
Deprecated==1.2.14
envier==0.5.1
google-api-core==2.17.1
google-auth==2.28.1
google-cloud-resource-manager==1.12.3
googleapis-common-protos==1.62.0
grpc-google-iam-v1==0.13.0
grpcio==1.62.0
grpcio-status==1.62.0
idna==3.6
importlib-metadata==6.11.0
opentelemetry-api==1.23.0
proto-plus==1.23.0
protobuf==4.25.3
pyasn1==0.5.1
pyasn1-modules==0.3.0
requests==2.31.0
rsa==4.9
setuptools==69.1.0
six==1.16.0
sqlparse==0.4.4
typing_extensions==4.10.0
urllib3==2.2.1
wheel==0.42.0
wrapt==1.16.0
xmltodict==0.13.0
zipp==3.17.0
```

</details>

### How can we reproduce your problem?

I've been able to reproduce the issue in a stripped down docker container. This will require you to retrieve a set of [GCP Application Default Credentials](https://cloud.google.com/docs/authentication/client-libraries#python). Save those credentials to a file called `key.json`.

Then save the following two files:

`dd_test.py`
```
import asyncio

from google.cloud.resourcemanager_v3 import ProjectsAsyncClient


async def main():
    client = ProjectsAsyncClient()

    await client.list_projects()


if __name__ == "__main__":
    asyncio.run(main())
```

`dd_test.Dockerfile`
```
FROM python:latest

RUN pip install google-cloud-resource-manager "ddtrace==2.7.1"

COPY key.json /key.json
COPY dd_test.py /dd_test.py

ENV GOOGLE_APPLICATION_CREDENTIALS=/key.json
ENV PYTHONFAULTHANDLER=1

CMD ["ddtrace-run", "python", "dd_test.py"]
```

Now run the following commands:
```
docker build -t dd-test -f dd_test.Dockerfile .
docker run --rm dd-test
```

### What is the result that you get?

A Segmentation Fault. The stack trace has so far been unhelpful, but here it is:
```
❯ docker run --rm dd-test
Fatal Python error: Segmentation fault

Current thread 0x0000ffffa8caf180 (most recent call first):
  File "/usr/local/lib/python3.12/threading.py", line 1010 in run
  File "/usr/local/lib/python3.12/threading.py", line 1073 in _bootstrap_inner
  File "/usr/local/lib/python3.12/threading.py", line 1030 in _bootstrap

Thread 0x0000ffffb8907180 (most recent call first):
  File "/usr/local/lib/python3.12/threading.py", line 359 in wait
  File "/usr/local/lib/python3.12/threading.py", line 655 in wait
  File "/usr/local/lib/python3.12/site-packages/ddtrace/internal/periodic.py", line 55 in run
  File "/usr/local/lib/python3.12/threading.py", line 1073 in _bootstrap_inner
  File "/usr/local/lib/python3.12/threading.py", line 1030 in _bootstrap

Thread 0x0000ffffb91ab180 (most recent call first):
  File "/usr/local/lib/python3.12/threading.py", line 359 in wait
  File "/usr/local/lib/python3.12/threading.py", line 655 in wait
  File "/usr/local/lib/python3.12/site-packages/ddtrace/internal/periodic.py", line 55 in run
  File "/usr/local/lib/python3.12/threading.py", line 1073 in _bootstrap_inner
  File "/usr/local/lib/python3.12/threading.py", line 1030 in _bootstrap

Thread 0x0000ffffb9a4f180 (most recent call first):
  File "/usr/local/lib/python3.12/threading.py", line 359 in wait
  File "/usr/local/lib/python3.12/threading.py", line 655 in wait
  File "/usr/local/lib/python3.12/site-packages/ddtrace/internal/periodic.py", line 55 in run
  File "/usr/local/lib/python3.12/threading.py", line 1073 in _bootstrap_inner
  File "/usr/local/lib/python3.12/threading.py", line 1030 in _bootstrap

Thread 0x0000ffffbaecf180 (most recent call first):
  File "/usr/local/lib/python3.12/threading.py", line 359 in wait
  File "/usr/local/lib/python3.12/threading.py", line 655 in wait
  File "/usr/local/lib/python3.12/site-packages/ddtrace/internal/periodic.py", line 55 in run
  File "/usr/local/lib/python3.12/threading.py", line 1073 in _bootstrap_inner
  File "/usr/local/lib/python3.12/threading.py", line 1030 in _bootstrap

Thread 0x0000ffffbce88020 (most recent call first):
  File "/usr/local/lib/python3.12/selectors.py", line 468 in select
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1947 in _run_once
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 639 in run_forever
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 672 in run_until_complete
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 118 in run
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 194 in run
  File "//dd_test.py", line 13 in <module>

Extension modules: ddtrace.vendor.psutil._psutil_linux, ddtrace.vendor.psutil._psutil_posix, ddtrace.internal._encoding, ddtrace.internal._rand, ddtrace.internal._tagset, google._upb._message, grpc._cython.cygrpc, charset_normalizer.md (total: 8)
```

### What is the result that you expected?

No segmentation fault.

**Comment by emmettbutler** [comment]
Thanks for including a minimal reproduction case, @sstoops. We'll check it out.

**Comment by dummerbd** [comment]
I'm also running into this issue with `google-api-core==2.18.0`. I can confirm that disabling the `grpc` patch eliminated the issue.

**Comment by dvgica** [comment]
I can also confirm all the details above, looking forward to a published version that includes the fix.

**Comment by lj-wego** [comment]
Hi, I'm also experiencing segfaults, using dd-trace `2.8.3` and google-api-core `2.18.0`

The segfault doesn't happen when disabling the grpc patch

**Comment by wconti27** [comment]
Hi all, the fix for GRPC tracing was released with `2.8.5`. To receive tracing with grpc, please re-enable the grpc patch with `_DD_TRACE_GRPC_AIO_ENABLED=true`
